pysmt/oracles.py

```diff
@@ -191,12 +191,13 @@ def walk_combine(self, formula, args, **kwargs):
 
     @walkers.handles(op.REAL_CONSTANT, op.BOOL_CONSTANT)
     @walkers.handles(op.INT_CONSTANT, op.BV_CONSTANT)
+    @walkers.handles(op.ALGEBRAIC_CONSTANT)
     @walkers.handles(op.STR_CONSTANT)
     def walk_constant(self, formula, args, **kwargs):
         """Returns a new theory object with the type of the constant."""
         #pylint: disable=unused-argument
         theory_out = Theory()
-        if formula.is_real_constant():
+        if formula.is_real_constant() or formula.is_algebraic_constant():
```

Comment:
I wonder if we want to distinguish this case. For example, for Arrays, we have the default constant value of arrays that is supported by MathSAT and Z3 (but not SMT-LIB), so we distinguish it explicitly in the logic https://github.com/pysmt/pysmt/blob/d1fadc09e745c7286abad93ba51e72cf6e285f58/pysmt/logics.py#L43 .

This would then allow you to cleanly state that only z3 supports this extension, and prevent pysmt from trying to use any of the other solvers if an algebraic constant is present.

Reply:
I would suggest to distinguish explicitly in the logic only the case in which z3 numerals are used to represent irrational values.

One of the things I like about pysmt is the ability to use different smt solvers in the same application with very little effort. You can use one SMT solver for a certain part of the computation, take its output and feed it to another SMT solver to continue the computation. 
In my case I would like, for instance, to take whatever output or model I get from z3 and use it to compute some SMT-query I submit to mathsat. If z3 in the model returns some algebraic constant I would like pysmt to take care of that when converting the expression for mathsat as much as possible.   

That said, I agree that carrying around a structure specific to z3 is not the cleanest way to achieve this.
A possibly cleaner way could be to immediately convert any z3 Numeral representing an integer or rational value in the output of z3 to a pysmt fraction (can we achieve this in te backward converter?). While irrational values could be carried around as algebraic values, associating such expressions with the "specific" logic.     
However, irrational constants like `sqrt(2)` that might be represented as z3 Numeral could also be handled by any other solver supporting `sqrt` as operator. The solution above will not handle these cases, but I think it could be acceptable anyway.